localstack/plugin/manager.py

```diff
@@ -0,0 +1,316 @@
+import logging
+import threading
+from typing import Any, Callable, Dict, Generic, List, Tuple, TypeVar, Union
+
+from .core import Plugin, PluginFinder, PluginLifecycleListener, PluginSpec, PluginSpecResolver
+
+LOG = logging.getLogger(__name__)
+
+P = TypeVar("P", bound=Plugin)
+
+
+def _call_safe(func: Callable, args: Tuple, exception_message: str):
+    """
+    Call the given function with the given arguments, and if it fails, log the given exception_message.
+    If loggin.DEBUG is set for the logger, then we also log the traceback.
+
+    :param func: function to call
+    :param args: arguments to pass
+    :param exception_message: message to log on exception
+    :return: whatever the func returns
+    """
+    try:
+        return func(*args)
+    except Exception as e:
+        if LOG.isEnabledFor(logging.DEBUG):
+            LOG.exception(exception_message)
+        else:
+            LOG.error("%s: %s", exception_message, e)
+
+
+class PluginLifecycleNotifierMixin:
```

Comment:
This class could be implemented in a generic way by implementing a decorator.
Something like this (untested!):
```
class SafeCall(object):
    def __init__(self,wrapped_class,*args,**kargs):
        self.wrapped_class = wrapped_class(*args,**kargs)

    def __getattr__(self,attr):
        orig_attr = self.wrapped_class.__getattribute__(attr)
        if callable(orig_attr):
            def wrapped(*args, **kwargs):
                try:
                    return orig_attr(*args, **kwargs)
                except Exception as e:
                    exception_message = f"error while calling {attr}"
                    if LOG.isEnabledFor(logging.DEBUG):
                        LOG.exception(exception_message)
                    else:
                        LOG.error("%s: %s", exception_message, e)
            return wrapped
        else:
            return orig_attr
```

You could then remove the `PluginLifecycleNotifierMixIn` completely and wrap any given (or the default) `PluginLifecycleListener` with the SafeCall wrapper (on `manager.py#132`).

Reply:
that's a pretty cool decorator :D after thinking about it i would rather have this implicitly in the PluginManager than rely on users to explicitly use the decorator in their implementation. since the code is relatively stable, i think it's fairly maintainable the way it is now.